### DISCUSSION TITLE
After the collision, the collision body determines the speed, but before it starts moving, is it possible to change its direction of movement?

### DISCUSSION THREAD

**Discussion by SourceCoder-2017** [bepu/bepuphysics1#54] [Q&A]
Hello, I have a question. I want to change the speed and direction of the collision body after two objects collide and contact, and determine the speed at the moment of collision, but before they start moving. Is there such an interface? Is it in InitialCollisionDetected? Is this callback after the collision speed is determined but before the position changes? If I change the speed of the object in this callback, will there be other simulation processes to change the speed of the object later?

**Comment by RossNordby** [comment]
Present-tense events like DetectingInitialCollision would give you a callback from within the execution of collision detection. Past-tense events like InitialCollisionDetected are dispatched in a deferred way at the end of the frame. (The reason for the split is that the 'inline' version of events requires that you deal with potentially multithreaded dispatch, since they're called from within collision detection.)

If you do use the inline events, note that modifying the velocity of a body is a potentially dangerous operation. Multiple events might fire simultaneously across different threads and attempt to modify the same body values. _Reading_ the velocity from multiple threads, though, would be fine.

**Reply by SourceCoder-2017** [reply]
I modified the moving direction of the collided object in DetectingInitialCollision, but it didn't seem to work. The speed direction printed later is not the value I set. How can I ensure that the speed direction after the collision is the value I set?

**Reply by SourceCoder-2017** [reply]
My own test shows that I need to keep changing the direction in the DetectingInitialCollision, PairTouching, and CollisionEnding methods to ensure that the velocity direction of the collided object is the corrected value after the collision ends.

**Reply by RossNordby** [reply]
There will be at least one (and likely more than one) executions of the solver between receiving a `DetectingInitialCollision` and `CollisionEnding`. The solver will adjust velocities to prevent and reduce penetration between bodies.

If you want complete control over velocity, it sounds like you want [`CollisionRule.NoSolver`.](https://github.com/bepu/bepuphysics1/blob/master/Documentation/CollisionRules.md)
